File: Assignment 2/Lab 2 Assignment/presentation.py
from directed_graph import write_graph
from errors import GraphException


class UI:
    def __init__(self, graph):
        self.__graph = graph
        self.__commands = {
            '1': self.__ui_get_number_of_vertices,
            '2': self.__ui_get_number_of_edges,
            '3': self.__ui_get_set_of_vertices,
            '4': self.__ui_get_set_of_edges,
            '5': self.__ui_check_vertex,
            '6': self.__ui_check_edge,
            '7': self.__ui_get_cost_of_edge,
            '8': self.__ui_get_in_degree,
            '9': self.__ui_get_out_degree,
            '10': self.__ui_get_outbound_neighbours,
            '11': self.__ui_get_inbound_neighbours,
            '12': self.__ui_change_edge_cost,
            '13': self.__ui_add_edge,
            '14': self.__ui_remove_edge,
            '15': self.__ui_add_vertex,
            '16': self.__ui_remove_vertex,
            '17': self.__ui_save_graph_to_file,
            '18': self.__ui_lowest_length_path,
            '19': self.__ui_strongly_connected_components,
        }

    @staticmethod
    def __print_menu():
        print("Graph operations:\n"
              "\t1 - Get the number of vertices in the graph\n"
              "\t2 - Get the number of edges in the graph\n"
              "\t3 - Get the set of vertices in the graph\n"
              "\t4 - Get the set of edges in the graph\n"
              "\t5 - Check if a given vertex is in the graph\n"
              "\t6 - Check if there is an edge between 2 vertices\n"
              "\t7 - Get the cost of an edge\n"
              "\t8 - Get the in degree of a vertex\n"
              "\t9 - Get the out degree of a vertex\n"
              "\t10 - Get the outbound neighbours of a vertex\n"
              "\t11 - Get the inbound neighbours of a vertex\n"
              "\t12 - Modify the cost of an edge\n"
              "\t13 - Add an edge\n"
              "\t14 - Remove an edge\n"
              "\t15 - Add a vertex\n"
              "\t16 - Remove a vertex\n"
              "\t17 - Write the graph to a file\n"
              "\t**18 - Find a lowest length path between 2 vertices\n"
              "\t**19 - Find the strongly connected components of the graph\n"
              "\tx - Exit\n"
              "** - Added in Assignment 2")

    def __ui_get_number_of_vertices(self):
        print(f"The number of vertices in the graph is {self.__graph.get_no_vertices()}.")

    def __ui_get_number_of_edges(self):
        print(f"The number of edges in the graph is {self.__graph.get_no_edges()}.")

    def __ui_get_set_of_vertices(self):
        print(f"The set of vertices in the graph is: ", end='')
        print(', '.join(str(vertex) for vertex in self.__graph.get_all_vertices()))

    def __ui_get_set_of_edges(self):
        print(f"The set of edges in the graph is:")
        for _from, _to, _cost in sorted(self.__graph.get_all_edges()):
            print(f"{_from}->{_to}, cost: {_cost}")

    def __ui_check_vertex(self):
        vertex = int(input("Give the vertex you want to check: "))
        if self.__graph.is_vertex_in_graph(vertex):
            print(f"Yes, the vertex {vertex} is in the graph.")
        else:
            print(f"No, the vertex {vertex} is not in the graph.")

    def __ui_check_edge(self):
        _from = int(input("Give the starting edge: "))
        _to = int(input("Give the ending edge: "))
        if self.__graph.is_edge_in_graph(_from, _to):
            cost = self.__graph.get_cost_of_edge(_from, _to)
            print(f"Yes, there is an edge between {_from} and {_to} (with a cost of {cost})")
        else:
            print(f"No, there is no edge between {_from} and {_to}.")

    def __ui_get_cost_of_edge(self):
        _from = int(input("Give the starting edge: "))
        _to = int(input("Give the ending edge: "))
        print(f"The cost of the edge {_from}->{_to} is {self.__graph.get_cost_of_edge(_from, _to)}.")

    def __ui_get_in_degree(self):
        vertex = int(input("Give the vertex: "))
        print(f"The in degree of the vertex {vertex} is {self.__graph.get_in_degree(vertex)}.")

    def __ui_get_out_degree(self):
        vertex = int(input("Give the vertex: "))
        print(f"The out degree of the vertex {vertex} is {self.__graph.get_out_degree(vertex)}.")

    def __ui_get_outbound_neighbours(self):
        vertex = int(input("Give the vertex: "))
        if self.__graph.get_out_degree(vertex) == 0:
            print(f"The vertex {vertex} has no outbound neighbours.")
            return
        print(f"The outbound neighbours of vertex {vertex} are:")
        for neighbour in sorted(self.__graph.get_outbound_neighbours(vertex)):
            print(f"\t{neighbour} (cost of {self.__graph.get_cost_of_edge(vertex, neighbour)})")

    def __ui_get_inbound_neighbours(self):
        vertex = int(input("Give the vertex: "))
        if self.__graph.get_in_degree(vertex) == 0:
            print(f"The vertex {vertex} has no outbound neighbours.")
            return
        print(f"The inbound neighbours of vertex {vertex} are:")
        for neighbour in sorted(self.__graph.get_inbound_neighbours(vertex)):
            print(f"\t{neighbour} (cost of {self.__graph.get_cost_of_edge(neighbour, vertex)})")

    def __ui_change_edge_cost(self):
        _from = int(input("Give the starting vertex: "))
        _to = int(input("Give the ending vertex: "))
        new_cost = int(input("Give the new cost of the edge: "))
        self.__graph.change_edge_cost(_from, _to, new_cost)
        print(f"The cost of the edge {_from}->{_to} was modified to {new_cost}.")

    def __ui_add_edge(self):
        _from = int(input("Give the starting vertex: "))
        _to = int(input("Give the ending vertex: "))
        cost = int(input("Give the cost of the edge: "))
        self.__graph.add_edge(_from, _to, cost)
        print(f"Edge {_from}->{_to} with a cost of {cost} added.")

    def __ui_remove_edge(self):
        _from = int(input("Give the starting vertex: "))
        _to = int(input("Give the ending vertex: "))
        self.__graph.remove_edge(_from, _to)
        print(f"Edge {_from}->{_to} removed.")

    def __ui_add_vertex(self):
        vertex = int(input("Give the new vertex: "))
        self.__graph.add_vertex(vertex)
        print(f"Vertex {vertex} added.")

    def __ui_remove_vertex(self):
        vertex = int(input("Give the vertex: "))
        self.__graph.remove_vertex(vertex)
        print(f"Vertex {vertex} removed.")

    def __ui_save_graph_to_file(self):
        file_name = input("Give the name of the file (with extension): ")
        graph_copy = self.__graph.get_copy_of_graph()
        write_graph(graph_copy, file_name)
        print(f"The graph was saved to {file_name}")

    def __ui_lowest_length_path(self):
        start = int(input("Give the start vertex: "))
        end = int(input("Give the end vertex: "))
        path = self.__graph.lowest_length_path(start, end)
        print(f"The lowest length path between {start} and {end} is: ")
        print(" -> ".join([str(number) for number in path]))
        print(f"The length of the path is {len(path) - 1}.")

    def __ui_strongly_connected_components(self):
        strongly_connected_components = self.__graph.kosaraju()
        print("These are the strongly connected components of the graph:")
        for component in sorted(strongly_connected_components):
            print(sorted(component))

    def start(self):
        while True:
            self.__print_menu()
            cmd = input("Command: ")
            if cmd == 'x':
                break
            elif cmd in self.__commands.keys():
                try:
                    self.__commands[cmd]()
                except GraphException as ge:
                    print(str(ge))
                except ValueError:
                    print("Invalid numerical input.")
            else:
                print("Invalid command.")
            print()

File: Assignment 1/Directed Graph - C++/Lab 1 Assignment - C++ Bonus Implementation/presentation.cpp
//
// Created by VladB on 20-Mar-21.
//

#include <iostream>
#include <string>

#include "presentation.h"
#include "utils.h"
#include "exceptions.h"

UI::UI(DirectedGraph &graph): graph(graph) {}

void UI::ui_get_number_of_vertices() {
    int nr_vertices = this->graph.get_nr_vertices();
    std::cout << "\nThe number of vertices in the graph is " << nr_vertices << "\n";
}

void UI::ui_get_number_of_edges() {
    int nr_edges = this->graph.get_nr_edges();
    std::cout << "\nThe number of edges in the graph is " << nr_edges << "\n";
}

void UI::run_app() {
    int cmd_as_int;
    std::string cmd;
    bool run_app = true;
    while (run_app) {
        UI::print_menu();
        std::cout << "\nPlease give a command: ";
        std::cin >> cmd;
        cmd_as_int = atoi(cmd.c_str());
        try {
            switch (cmd_as_int) {
                case 1:
                    this->ui_get_number_of_vertices();
                    break;
                case 2:
                    this->ui_get_number_of_edges();
                    break;
                case 3:
                    this->ui_get_vertices();
                    break;
                case 4:
                    this->ui_get_edges();
                    break;
                case 5:
                    this->ui_check_vertex();
                    break;
                case 6:
                    this->ui_check_edge();
                    break;
                case 7:
                    this->ui_get_cost_of_edge();
                    break;
                case 8:
                    this->ui_in_degree_of_vertex();
                    break;
                case 9:
                    this->ui_out_degree_of_vertex();
                    break;
                case 10:
                    this->ui_outbound_neighbours();
                    break;
                case 11:
                    this->ui_inbound_neighbours();
                    break;
                case 12:
                    this->ui_change_edge_cost();
                    break;
                case 13:
                    this->ui_add_edge();
                    break;
                case 14:
                    this->ui_remove_edge();
                    break;
                case 15:
                    this->ui_add_vertex();
                    break;
                case 16:
                    this->ui_remove_vertex();
                    break;
                case 17:
                    this->ui_write_graph_to_file();
                    break;
                case 18:
                    run_app = false;
                    break;
                default:
                    std::cout << "\nInvalid command given.\n";
                    break;
            }
        }
        catch (GraphException& ge) {
            std::cout << ge.what();
        }
    }
}

void UI::print_menu() {
    std::cout << "\nGraph operations:\n"
                 "\t1 - Get the number of vertices in the graph\n"
                 "\t2 - Get the number of edges in the graph\n"
                 "\t3 - Get the set of vertices in the graph\n"
                 "\t4 - Get the set of edges in the graph\n"
                 "\t5 - Check if a given vertex is in the graph\n"
                 "\t6 - Check if there is an edge between 2 vertices\n"
                 "\t7 - Get the cost of an edge\n"
                 "\t8 - Get the in degree of a vertex\n"
                 "\t9 - Get the out degree of a vertex\n"
                 "\t10 - Get the outbound neighbours of a vertex\n"
                 "\t11 - Get the inbound neighbours of a vertex\n"
                 "\t12 - Modify the cost of an edge\n"
                 "\t13 - Add an edge\n"
                 "\t14 - Remove an edge\n"
                 "\t15 - Add a vertex\n"
                 "\t16 - Remove a vertex\n"
                 "\t17 - Write the graph to a file\n"
                 "\t18 - Exit\n";
}

void UI::ui_get_vertices() {
    if (this->graph.get_nr_vertices() == 0) {
        std::cout << "\nThere are no vertices in the graph.\n";
        return;
    }
    std::pair<std::map<int, std::set<int>>::iterator, std::map<int, std::set<int>>::iterator> start_end_it;
    start_end_it = this->graph.out_relations();
    auto start_it = start_end_it.first;
    auto end_it = start_end_it.second;
    std::cout << "\nThese are the vertices from the graph: ";
    for (auto it = start_it; it != end_it; it++) {
        std::cout << it->first << " ";
    }
}

void UI::ui_get_edges() {
    if (this->graph.get_nr_edges() == 0) {
        std::cout << "\nThe graph has no edges.\n";
        return;
    }
    std::pair<std::map<std::pair<int, int>, int>::iterator, std::map<std::pair<int, int>, int>::iterator> start_end_it;
    start_end_it = this->graph.get_all_edges();
    auto start_it = start_end_it.first;
    auto end_it = start_end_it.second;
    std::cout << "\nThese are all the edges from the graph:\n";
    for (auto it = start_it; it != end_it; it++) {
        std::cout << it->first.first << "->" << it->first.second << ", cost of " << it->second << "\n";
    }
}

void UI::ui_check_vertex() {
    int vertex;
    std::cout << "\nGive the vertex: ";
    std::cin >> vertex;
    if (this->graph.is_vertex_in_graph(vertex)) {
        std::cout << "\nYes, the vertex " << vertex << " is in the graph.\n";
    }
    else {
        std::cout << "\nNo, the vertex " << vertex << " is NOT in the graph.\n";
    }
}

void UI::ui_check_edge() {
    int from, to;
    std::cout << "\nGive the starting vertex of the graph: ";
    std::cin >> from;
    std::cout << "\nGive the ending vertex of the graph: ";
    std::cin >> to;
    if (this->graph.is_edge_in_graph(from, to)) {
        int cost = this->graph.get_cost_of_edge(from, to);
        std::cout << "\nYes, the edge " << from << "->" << to << " is in the graph(cost of " << cost << ")\n";
    }else {
        std::cout << "\nNo, the edge " << from << "->" << to << " is NOT in the graph\n";
    }
}

void UI::ui_get_cost_of_edge() {
    int from, to;
    std::cout << "\nGive the starting vertex of the edge: ";
    std::cin >> from;
    std::cout << "\nGive the ending vertex of the edge: ";
    std::cin >> to;
    int cost = this->graph.get_cost_of_edge(from, to);
    std::cout << "\nThe cost of the edge " << from << "->" << to << "is " << cost << "\n";
}

void UI::ui_in_degree_of_vertex() {
    int vertex;
    std::cout << "\nGive the vertex number: ";
    std::cin >> vertex;
    int in_degree = this->graph.get_in_degree(vertex);
    std::cout << "\nThe in-degree of the vertex " << vertex << " is " << in_degree << "\n";
}

void UI::ui_out_degree_of_vertex() {
    int vertex;
    std::cout << "\nGive the vertex number: ";
    std::cin >> vertex;
    int out_degree = this->graph.get_out_degree(vertex);
    std::cout << "\nThe out-degree of the vertex " << vertex << " is " << out_degree << "\n";
}

void UI::ui_outbound_neighbours() {
    int vertex;
    std::cout << "\nGive the vertex number: ";
    std::cin >> vertex;
    if (this->graph.get_out_degree(vertex) == 0) {
        std::cout << "\nThe vertex " << vertex << " has no outbound neighbours.\n";
        return;
    }
    std::pair<std::set<int>::iterator, std::set<int>::iterator> start_end_it;
    start_end_it = this->graph.get_outbound_neighbours(vertex);
    auto start_it = start_end_it.first;
    auto end_it = start_end_it.second;
    int neighbour, cost;
    std::cout << "\nThe outbound neighbours of vertex " << vertex << " are:\n";
    for (auto it = start_it; it != end_it; it++) {
        neighbour = *it;
        cost = this->graph.get_cost_of_edge(vertex, neighbour);
        std::cout << neighbour << " (cost of " << cost << ")\n";
    }
}

void UI::ui_inbound_neighbours() {
    int vertex;
    std::cout << "\nGive the vertex number: ";
    std::cin >> vertex;
    if (this->graph.get_in_degree(vertex) == 0) {
        std::cout << "\nThe vertex " << vertex << " has no inbound neighbours.\n";
        return;
    }
    std::pair<std::set<int>::iterator, std::set<int>::iterator> start_end_it;
    start_end_it = this->graph.get_inbound_neighbours(vertex);
    auto start_it = start_end_it.first;
    auto end_it = start_end_it.second;
    int neighbour, cost;
    std::cout << "\nThe inbound neighbours of vertex " << vertex << " are:\n";
    for (auto it = start_it; it != end_it; it++) {
        neighbour = *it;
        cost = this->graph.get_cost_of_edge(neighbour, vertex);
        std::cout << neighbour << " (cost of " << cost << ")\n";
    }
}

void UI::ui_change_edge_cost() {
    int from, to, new_cost;
    std::cout << "\nGive the starting vertex of the edge: ";
    std::cin >> from;
    std::cout << "\nGive the ending vertex of the edge: ";
    std::cin >> to;
    std::cout << "\nGive the new cost of the edge: ";
    std::cin >> new_cost;
    this->graph.change_cost(from, to, new_cost);
    std::cout << "\nThe cost of the edge " << from << "->" << to << " was changed to " << new_cost << "\n";
}

void UI::ui_add_edge() {
    int from, to, cost;
    std::cout << "\nGive the starting vertex of the new edge: ";
    std::cin >> from;
    std::cout << "\nGive the ending vertex of the new edge: ";
    std::cin >> to;
    std::cout << "\nGive the cost of the new edge: ";
    std::cin >> cost;
    this->graph.add_edge(from, to, cost);
    std::cout << "\nThe edge " << from << "->" << to << "(cost of " << cost << ") was added to the graph.\n";
}

void UI::ui_remove_edge() {
    int from, to;
    std::cout << "\nGive the starting vertex of the edge: ";
    std::cin >> from;
    std::cout << "\nGive the ending vertex of the edge: ";
    std::cin >> to;
    this->graph.remove_edge(from, to);
    std::cout << "\nThe edge was removed.\n";
}

void UI::ui_add_vertex() {
    int vertex;
    std::cout << "\nGive the number of the new vertex: ";
    std::cin >> vertex;
    this->graph.add_vertex(vertex);
    std::cout << "\nThe vertex " << vertex << " was added to the graph.\n";
}

void UI::ui_remove_vertex() {
    int vertex;
    std::cout << "\nGive the number of the vertex: ";
    std::cin >> vertex;
    this->graph.remove_vertex(vertex);
    std::cout << "\nThe vertex " << vertex << " was removed from the graph.\n";
}

void UI::ui_write_graph_to_file() {
    char file_name[256];
    std::cout << "\nGive the name of the file (with extension): ";
    getchar();
    std::cin.getline(file_name, sizeof(file_name));
    write_graph_to_file(file_name, this->graph);
    std::cout << "\nThe graph was saved in " << file_name << "\n";
}




